sopel/irc/isupport.py

```diff
@@ -0,0 +1,380 @@
+# coding=utf-8
+"""IRC Tools for ISUPPORT management.
+
+When a server wants to advertise its features and settings, it can use the
+``RPL_ISUPPORT`` command with a list of arguments.
+
+.. seealso::
+
+    https://modern.ircdocs.horse/#rplisupport-005
+
+"""
+# Copyright 2019, Florian Strzelecki <florian.strzelecki@gmail.com>
+#
+# Licensed under the Eiffel Forum License 2.
+from __future__ import unicode_literals, absolute_import, print_function, division
+
+import functools
+import itertools
+import re
+
+
+def _optional(parser, default=None):
+    # set a parser as optional: will always return the default value provided
+    # if there is no value (empty or None)
+    @functools.wraps(parser)
+    def wrapped(value):
+        if not value:
+            return default
+        return parser(value)
+    return wrapped
+
+
+def _no_value(value):
+    # always ignore the value
+    return None
+
+
+def _single_character(value):
+    if len(value) > 1:
+        raise ValueError('Too many characters: %r.' % value)
+
+    return value
+
+
+def _map_items(parser=str, map_separator=',', item_separator=':'):
+    @functools.wraps(parser)
+    def wrapped(value):
+        items = sorted(
+            item.split(item_separator)
+            for item in value.split(map_separator))
+
+        return tuple(
+            (k, parser(v) if v else None)
+            for k, v in items
+        )
+    return wrapped
+
+
+def _parse_chanmodes(value):
+    items = value.split(',')
+
+    if len(items) < 4:
+        raise ValueError('Not enough channel types to unpack from %r.' % value)
+
+    # add extra channel type's modes to their own tuple
+    # result in (A, B, C, D, (E, F, G, H, ..., Z))
+    # where A, B, C, D = result[:4]
+    # and extras = result[4]
+    return tuple(items[:4]) + (tuple(items[4:]),)
+
+
+def _parse_elist(value):
+    # letters are case-insensitives
+    return tuple(sorted(set(letter.upper() for letter in value)))
+
+
+def _parse_extban(value):
+    args = value.split(',')
+
+    if len(args) < 2:
+        raise ValueError('Invalid value for EXTBAN: %r.' % value)
+
+    prefix = args[0] or None
+    items = tuple(sorted(set(args[1])))
+
+    return (prefix, items)
+
+
+def _parse_prefix(value):
+    result = re.match(r'\((?P<modes>\S+)\)(?P<prefixes>\S+)', value)
+
+    if not result:
+        raise ValueError('Invalid value for PREFIX: %r' % value)
+
+    modes = result.group('modes')
+    prefixes = result.group('prefixes')
+
+    if len(modes) != len(prefixes):
+        raise ValueError('Mode list does not match for PREFIX: %r' % value)
+
+    return tuple(sorted(zip(modes, prefixes)))
+
+
+ISUPPORT_PARSER = {
```

Comment:
Causes ripple-effect changes elsewhere, I'm aware. If you don't want to do this, meh. I'm just suggesting.

```suggestion
ISUPPORT_PARSERS = {
```